src/utils/generator.js

```diff
@@ -1253,3 +1285,113 @@ export class PreludeGenerator {
     return ref;
   }
 }
+
+// This function attempts to optimize Object.assign calls, by merging mulitple
+// calls into one another where possible. For example:
+//
+// var a = Object.assign({}, someAbstact);
+// var b = Object.assign({}, a);
+//
+// Becomes:
+// var b = Object.assign({}, someAbstract, a);
+//
+export function attemptToMergeEquivalentObjectAssigns(
+  realm: Realm,
+  callbacks: VisitEntryCallbacks,
+  temporalBuildNodeEntry: TemporalBuildNodeEntry
+): TemporalBuildNodeEntryOptimizationStatus | TemporalBuildNodeEntry {
+  let args = temporalBuildNodeEntry.args;
+  // If we are Object.assigning 3 or more args
+  if (args.length < 3) {
+    return "NO_OPTIMIZATION";
+  }
+  let objectAssingFunc = args[0];
+  let to = args[1];
+  // Then scan through the args after the "to" of this Object.assign, to see if any
+  // other sources are the "to" of a previous Object.assign call
+  for (let i = 2; i < args.length; i++) {
+    let possibleOtherObjectAssignTo = args[i];
+    // Ensure that the "to" value can be omitted
+    if (!callbacks.canOmit(possibleOtherObjectAssignTo)) {
+      continue;
+    }
+    // Check if the "to" was definitely an Object.assign, it should
+    // be a snapshot AbstractObjectValue
+    if (possibleOtherObjectAssignTo instanceof AbstractObjectValue) {
+      let otherTemporalBuildNodeEntry = realm.getTemporalBuildNodeEntryArgsFromDerivedValue(
+        possibleOtherObjectAssignTo
+      );
+      if (otherTemporalBuildNodeEntry === undefined) {
+        continue;
+      }
+      let otherArgs = otherTemporalBuildNodeEntry.args;
+      // Object.assign has at least 2 args and the first is that of Object.assign native function
+      if (otherArgs.length < 3 || otherArgs[0] !== objectAssingFunc) {
+        continue;
+      }
+      // If we cannot omit the "to" value that means it's being used, so we shall not try to
+      // optimize this Object.assign.
+      if (!callbacks.canOmit(to)) {
+        let newArgs = [objectAssingFunc, to];
```

Comment:
How do you know that the object referenced by `possibleOtherObjectAssignTo` hasn't been mutated by some other generator entries in between?

Reply:
We check a few lines up:

```js
    if (!callbacks.canOmit(possibleOtherObjectAssignTo)) {
      continue;
    }
```

Reply:
So you skip `possibleOtherObjectAssignTo` if it has already been visited.
However, if it hasn't already been visited, doesn't mean it won't be visited for other reasons later. It doesn't work this way.

Reply:
Also, what about the objects that were used to create `possibleOtherObjectAssignTo`, they could have been mutated in between.

Reply:
I guess we're confusing how Object.assign works then? If a tree of Object.assigns, say:

A -> B -> C -> D

Even if A, B or C have not been visited, but they may be visited, D won't apply any optimizations until D itself has been visited. I added a test for this too:

I added a bunch of tests around this. Unless I'm completely missing something, or maybe you can give a test case that fails this?